### ISSUE TITLE
Batching does not work when concat op is present in the graph during serving

### ISSUE THREAD

**Issue by kikushah** [tensorflow/serving#1305]
Hi,

I trained a model using python 3.6 and tensorflow 1.13.

I am training a densenet. One of the ops that I use is concat(with tf.zeros), to double the size of my input tensor. The model is trained with batch size 1.

It trains well, and the graph is exported with tf.saved_model.builder.SavedModelBuilder and served with tensorflow_model_server.

I use a c++ client to send prediction requests to the served model, from localhost.

When I send requests of batch size 1, it works well.

When I enable batching and increase the max batch size to a value greater than 1, the concat op fails.

Upon debugging, I found that the requests are getting batched properly but there is a problem in the concat op.
`
gRPC call return code: 3: _MklConcatOp : Dimensions of inputs should match: shape[0][0]= 10 vs. shape[1][0] = 1 [[{{node global_Model/rn_first/concat}} = _MklConcatV2[N=2, T=DT_FLOAT, Tidx=DT_INT32, _kernel="MklOp", _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_global_Model/global_Model/PlaceholderWithDefault_0_0, global_Model/global_Model/PlaceholderWithDefault/input, global_Model_-1/ArgMax/dimension, DMT/_0, DMT/_1, DMT/_2)]]
`
While sending single requests, input tensor is of shape (1, 50). While sending synchronous batch requests, the input tensor shape is (10, 50). While sending asynchronous request, the shape is (1, 50) but 10 such requests.

I am not sure if I am doing anything wrong here. This seems to be a problem or bug in the serving batching.

**Comment by hgadig** [comment]
Hi @kikushah, could you please provide the following information which help us to look into the issue. 

### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **TensorFlow Serving installed from (source or binary)**:
- **TensorFlow Serving version**:

### Exact Steps to Reproduce
Please include steps followed and a reproducible code snippet for someone to reproduce this issue on their
own machine. Please provide 

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem.
If including tracebacks, please include the full traceback. Large logs and files
should be attached. Try to provide a reproducible test case that is the bare
minimum necessary to generate the problem.

**Comment by kikushah** [comment]
Hi,
I use Ubuntu 16.04.

TensorFlow serving is installed from source with --config=mkl 

TensorfFlow version is 1.31.1

concat op in python is built with the following code:

`
fcweightsinitializer=variance_scaling

def halfresnet(x, scope, nh, act=actfunction, init_scale=1.0, sublayers=1, collections=None):   
    with tf.variable_scope(scope):
        nin = x.get_shape()[1].value
        zeros = tf.zeros(x.get_shape(), dtype=x.dtype)
        w = tf.get_variable("w0", [nin, nh], initializer=fcweightsinitializer(init_scale), collections=collections)
        b = tf.get_variable("b0", [nh], initializer=tf.constant_initializer(0.0), collections=collections)
        z = tf.matmul(x, w) + b
        x = tf.concat(axis=1, values=[x, zeros])
        z = z + x
        h = act(z)
        return h`

This is enough to reproduce the problem. When this model is saved using 
`tf.saved_model.builder.SavedModelBuilder(savepath)
builder.add_meta_graph_and_variables(sess=get_session(sess),
                                             tags=[tf.saved_model.tag_constants.SERVING],
                                             signature_def_map=MyTFSignatureDefsMap,
                                             #legacy_init_op = legacy_init_op,
                                             clear_devices=True)
        builder.save()`

The serving client is written in c++ from where grpc requests are sent to the above model which is served using tensorflow_model_server with the enable batching flag on.

`
Tensor X(tensorflow::DT_Float, tensorflow::TensorShape({1, 50}))
// This Tensor is then filled with values from a different bit of code. 
PredictRequest predictRequest;
PredictResponse response;
ClientContext context;
model_signature_name = to_string(index);	
predictRequest.mutable_model_spec()->set_name(this->model_name);
PredictRequest.mutable_model_spec()->set_signature_name(this->model_signature_name);	
google::protobuf::Map<tensorflow::string, tensorflow::TensorProto>& inputs = *predictRequest.mutable_inputs();
tensorflow::TensorProto protoX;
X.AsProtoTensorContent(&protoX);
`
Logs:
gRPC call return code: 3: _MklConcatOp : Dimensions of inputs should match: shape[0][0]= 10 vs. shape[1][0] = 1 [[{{node global_Model/rn_first/concat}} = _MklConcatV2[N=2, T=DT_FLOAT, Tidx=DT_INT32, _kernel="MklOp", _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_global_Model/global_Model/PlaceholderWithDefault_0_0, global_Model/global_Model/PlaceholderWithDefault/input, global_Model_-1/ArgMax/dimension, DMT/_0, DMT/_1, DMT/_2)]]

**Comment by secsilm** [comment]
Same issue here. Python 3.6 and tf serving 1.12. 

If I use batch 1 or `saved_model_cli` or `tf.contrib.predictor.from_saved_model`, it all works fine. But if I use restful api and batch > 1, it'll raise an error about reshape op:

```bash
{'error': 'Input to reshape is a tensor with 256 values, but the requested shape has 128\n\t [[{{node bert/encoder/Reshape}} = Reshape[T=DT_INT32, Tshape=DT_INT32, _device="/job:localhost/replica:0/task:0/device:GPU:0"](_arg_input_mask_0_1/_423, bert/encoder/Reshape/shape)]]\n\t [[{{node loss/Softmax/_427}} = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_1841_loss/Softmax", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]'}
```

**Comment by kikushah** [comment]
@secsilm - agreed. It works well for multiple individual requests because serving is processing them serially and not batching them. Can you check the time taken to execute in the first case?

I have the same issue.

**Comment by workd91** [comment]
Hi! I think the error is because tensor x is assigned a shape of (1, 50). The model has to be changed for (None, 50) to be batched. Mismatch in the input happens because of that. Input is having (10, 50) while the other variable in the concatenate step is with (1, 50)

**Comment by secsilm** [comment]
@kikushah The cost time is about 20 ms in the first case.

**Comment by workd91** [comment]
Mismatch in the input happens. Input is having (10, 50) after batching in the abstraction layer while the other (zeros) variable in the concatenate step is with (1, 50)

from the tf.concat 
"The number of dimensions of the input tensors must match, and all dimensions except axis must be equal."

https://www.tensorflow.org/api_docs/python/tf/concat 

During the synchronous request, the tensorflow session is grouped. In this case to make the concat function work. Keep the batch size = 10. Zero pad if the batch size is less. 
step - 1 
- >  zeros variable zeroth dimension should set be equal to the batch size in this case 10. 
step -2 
You can create a batching config file -->  batch_config.txt with the following parameters 
max_batch_size { value: 10}
pad_variable_length_inputs: true
allowed_batch_sizes: 10

Step - 3 In the docker command line 
--enable_batching --batching_parameters_file=/models/batching_config.txt
Allow copy the batching_config.txt to docker models file. 
--mount type=bind, source='/home/path-to-file/batching_config.txt, target=/models/batching_config.txt

**Comment by kikushah** [comment]
@workd91 - In my training job, the eval model has a batch size 1 and that is the model that I am serving. If I change it batch size 10, will I not have to train again, because the dimensions of the weights and biases will have to change accordingly? and What happens if I want to increase the batch size to 100 after a few months? Shouldn't there be scalability without having to retrain?

**Comment by christisg** [comment]
@kikushah - as per your own investigation, serving batching is working as expected in this case, i.e the synchronous batch requests are getting grouped into the same batch and the result input tensor shape at runtime is (10, 50), which mismatches 'zeros' shape.
What might solve the issue is if you create 'zeros' using tf.shape(x) instead of x.get_shape(). The former is dynamic (true) shape computed at runtime, while the latter is static (inferred) shape.

Let us know if this is helpful.

**Comment by kikushah** [comment]
@christisg - Thanks for the response, it make sense.

I understand how zeros can be created using tf.shape(x)

How about these lines? 
`nin = x.get_shape()[1].value
w = tf.get_variable("w0", [nin, nh], initializer=fcweightsinitializer(init_scale), `collections=collections)`

Even they have a dependency on shape of x and am not really sure if the weights can be dynamically shaped.

Please correct me if I am wrong.

**Comment by alextp** [comment]
TF variables can have partially unspecified shapes. I prefer to construct those using `tf.Variable(initial_value, shape=[10, None, 5], use_resource=True)` as it's the way we'll recommend people to do it in tf2.0. What this means is that you can assign to this variable any value whose shape is compatible with the specified shape argument.

**Comment by rmothukuru** [comment]
@kikushah ,
Can you please let us know if @alextp's suggestion has resolved your issue.

**Comment by kikushah** [comment]
@alextp 's suggestion along with everybody's response helped.

You may close this issue.

**Comment by rmothukuru** [comment]
Closing this issue as it has been resolved.

**Comment by rmothukuru** [comment]
Are you satisfied with the resolution of your issue?<br> [Yes](https://docs.google.com/forms/d/e/1FAIpQLSfaP12TRhd9xSxjXZjcZFNXPGk4kc1-qMdv3gc6bEP90vY1ew/viewform?entry.85265664=Yes&entry.2137816233=1305)<br> [No](https://docs.google.com/forms/d/e/1FAIpQLSfaP12TRhd9xSxjXZjcZFNXPGk4kc1-qMdv3gc6bEP90vY1ew/viewform?entry.85265664=No&entry.2137816233=1305)
